tests/rados/test_osd_async_recovery.py

```diff
@@ -0,0 +1,143 @@
+"""
+This file contains the  methods to verify the async messages in the OSD logs.
+"""
+import time
+import traceback
+
+from ceph.ceph_admin import CephAdmin
+from ceph.rados.core_workflows import RadosOrchestrator
+from tests.rados.monitor_configurations import MonConfigMethods
+from tests.rados.stretch_cluster import wait_for_clean_pg_sets
+from utility.log import Log
+from utility.utils import method_should_succeed
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw):
+    """
+    CEPH-83574487 - Verification of Async recovery performance with fault injection.
+    1.Create pool and write data in to the pool
+    2.set the osd_async_recovery_min_cost to 10  and debug_osd values to 20
+    3.Start writing the data into pool and start scrub parallely
+    4.Once the scrub started check the OSD logs contain the-
+       - choose_async_recovery_replicated candidates by cost are
+       - choose_async_recovery_replicated
+    5. Once the tests are complete delete the pool and remove the set values of the parameter
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_object = RadosOrchestrator(node=cephadm)
+    mon_obj = MonConfigMethods(rados_obj=rados_object)
+    ceph_nodes = kw.get("ceph_nodes")
+    installer = ceph_cluster.get_nodes(role="installer")[0]
+    try:
+        target_configs = config["async_recovery"]["configurations"]
+        mon_obj.set_config(
+            section="osd", name="osd_async_recovery_min_cost", value="10"
+        )
+        mon_obj.set_config(section="osd", name="debug_osd", value="20")
+
+        # Creating pools and starting the test
+        for entry in target_configs.values():
+            log.debug(
+                f"Creating {entry['pool_type']} pool on the cluster with name {entry['pool_name']}"
+            )
+            if entry.get("pool_type", "replicated") == "erasure":
+                method_should_succeed(
+                    rados_object.create_erasure_pool, name=entry["pool_name"], **entry
+                )
+            else:
+                method_should_succeed(
+                    rados_object.create_pool,
+                    **entry,
+                )
+            rados_object.bench_write(
+                pool_name=entry["pool_name"], rados_write_duration=600, background=True
+            )
+
+            log_lines = [
+                "choose_async_recovery_replicated candidates by cost are",
+                "choose_async_recovery_replicated",
+            ]
+            osd_list = []
+            for node in ceph_nodes:
+                if node.role == "osd":
+                    node_osds = rados_object.collect_osd_daemon_ids(node)
```

Comment:
We can just pick 1 osd from each host for reboot

Reply:
I changed the logic so that it picks the one OSD from the node, and it failed most of the time. I am thinking of a scenario where, when data is written to OSDs, there is no guarantee that it needs to write the exact OSD, which I restated.In this scenario the messages does not exist on the OSD log.  If we pick one OSD from a node, there are very few chances of getting async messages in the OSD.

Logs: 
  1. http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-PELJ74
  2. http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-TU6Q4D
  3. 	http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-JLASOA